common_constants.py

```diff
@@ -1,18 +1,100 @@
 # Definition of 3SAT
 VARS_PER_CLAUSE = 3
 
-# common vectors
-ZERO = [0, 0]
-LEFT = [0, -1]
-RIGHT = [0, 1]
-UP = [1, 0]
-DOWN = [-1, 0]
-
 # colors named based on in-game functionality
 COLORS = {
           'ladder': (0, 0, 255),
           'hard': (210, 180, 140),
           'soft': (128, 128, 128),
           'princess': (250, 20, 200),
-          'support': (255, 255, 255)
-        }
\ No newline at end of file
+          'support': (255, 255, 255),
+          'air': (255, 255, 255),
+          'slab': (128, 128, 128),
+          'border': (0, 0, 0),
+          'tip':  (110,  110, 110),
+          'gripped': (50, 50, 50)
+        }
+
+class Vector:
+    def __init__(self, x, y):
+        self.x = x
+        self.y = y
+        from math import sqrt
+        self.magnitude = sqrt(self.x * self.x + self.y * self.y)
+
+    def __matmul__(self, other):
+        return self.x * other.x + self.y * other.y
+
+    # pure function
```

Comment:
...Meaning this is overriding an abstract method?

Reply:
no, pure in the functional programming sense of having no side effects